### PR TITLE
Storage Component

### PR DESCRIPTION
**Pull request by ccoffrin**
Adding support for a generic storage component.

Feedback welcome, will keep open for 48 hours.

CC @frederikgeth, @hakanergun, @rb004f, @harshangrjn, @hhijazi, @kaarthiksundar, @bluejuniper, @claytonpbarrows, @jd-lara

### CODE REVIEW

File: src/core/data.jl [hunk 1]
```diff
@@ -717,6 +748,74 @@ function check_transformer_parameters(data::Dict{String,Any})
...
+
+        if strg["discharge_efficiency"] < 0.0 || strg["discharge_efficiency"] > 1.0
+            warn(LOGGER, "storage unit $(strg["index"]) discharge efficiency of $(strg["eff_discharge"]) is out of the valid range 0.0 to 1.0)")
+        end
+
+        if !isapprox(strg["x"], 0.0, atol=1e-6, rtol=1e-6)
+            warn(LOGGER, "storage unit $(strg["index"]) has a non-zero reactance $(strg["x"]), which is currently ignored")
+        end
+
+        if strg["charge_rating"] > strg["energy_rating"]
```
Comment (frederikgeth) [inline src/core/data.jl]: the charge rating only limits the charge power directly. the interaction with the energy rating depends on the duration of the time steps. e.g. 5 kW charging over at 80% efficiency results in the energy stored increasing by 1 kWh every 0.25 h or 4 kWh every hour. I suggest to either drop this or validate that time_elapsed is strictly larger than 0.

File: src/core/data.jl [hunk 2]
```diff
@@ -717,6 +748,74 @@ function check_transformer_parameters(data::Dict{String,Any})
...
+        end
+        if haskey(strg, "current_rating") && strg["current_rating"] < 0.0
+            error(LOGGER, "storage unit $(strg["index"]) has a non-positive current rating $(strg["thermal_rating"])")
+        end
+
+        if strg["charge_efficiency"] < 0.0 || strg["charge_efficiency"] > 1.0
+            warn(LOGGER, "storage unit $(strg["index"]) charge efficiency of $(strg["eff_charge"]) is out of the valid range 0.0 to 1.0)")
+        end
+
+        if strg["discharge_efficiency"] < 0.0 || strg["discharge_efficiency"] > 1.0
```
[the remark this answers is not in the document]
Comment (ccoffrin) [reply src/core/data.jl]: Good point. Done.

File: src/core/data.jl (hunk 1 above)
Comment (ccoffrin) [reply src/core/data.jl]: Good point, I removed these.

**Comment by jd-lara** [comment]
I would suggest for clarity in the data model to name the `thermal_rating` and `current_rating` to mention this is the converter ratings such that they distinct from the storage ratings `charge_rating` and
   `discharge_rating`.

**Pull request merged**